buzzmobile/tests/test_utils/mock_nodes.py

```diff
@@ -4,114 +4,112 @@
-import rospy
+from StringIO import StringIO
+
+import rosnode
+
 
 class TimeoutError(Exception):
     """Py3 shim, represents no response from a syscall.
     """
     pass
 
 
-
 class NoMessage(Exception):
     """Exception for a lack of message in a Node.
     """
     pass
 
-class MockNode(object):
+class MockPublisher(object):
     """Mock of a node object for testing.
     """
-    def __init__(self, topic, msg_type, node):
+    def __init__(self, topic, msg_type, queue_size):
         self.topic = topic
         self.msg_type = msg_type
-        self.node = node
+        pub_data = pickle.dumps((topic, msg_type, queue_size))
+        location = './tests/test_utils/publisher.py'
+        self.proc = subprocess.Popen([location, pub_data],
+                stdin=subprocess.PIPE)
 
     def send(self, value):
         """Sends data to be publoshed to the mocked topic.
         """
-        self.node.publish(value)
+        value.serialize(self.proc.stdin)
         # Just long enough to prevent out of order
-        time.sleep(.1)
+        time.sleep(.2)
+
+    def kill(self):
+        """Kills the publisher ros node process.
+        """
+        self.proc.kill()
+
 
 @contextlib.contextmanager
 def mock_pub(topic, rosmsg_type, queue_size=1):
     """Mocks a node and cleans it up when done.
     """
-    pub = rospy.Publisher(topic, rosmsg_type, queue_size=queue_size)
-    yield MockNode(topic, rosmsg_type, pub)
-    pub.unregister()
+    pub = MockPublisher(topic, rosmsg_type, queue_size)
+    no_ns = topic.split('/')[-1]
+    while not any(nn.split('/')[-1].startswith(
+        ''.join(['mock_publish_', no_ns])) for nn in rosnode.get_node_names()):
+        time.sleep(.1)
+    try:
+        yield pub
+    finally:
+        pub.kill()
 
 
-class TestNode(object):
+class MockListener(object):
     """Wrapper around a node used for testing.
     """
 
     def __init__(self, topic, msg_type):
         self.topic = topic
         self.msg_type = msg_type
-        self.received = False
+        # get port and rosmaster uri cleanly
+
+        # do this better, somehow!
+        location = './tests/test_utils/listener.py'
+        self.proc = subprocess.Popen([location,
+            pickle.dumps((topic, msg_type))], stdout=subprocess.PIPE)
         self._message = None
 
+    def kill(self):
+        """Kills the subscriber ros node process.
+        """
+        self.proc.kill()
+
     @property
     def message(self):
         """Getter for the message property.
 
         Makes sure you've actually recieved a message before providing one.
         """
-        if not self.received:
-            raise NoMessage(('No message has been '
-                'published to {} yet').format(self.topic))
+        if  not self._message:
+            msg = self.msg_type()
+            sio = StringIO()
+            msg.serialize(sio)
+            data = self.proc.stdout.read(sio.len)
+            msg.deserialize(data)
+            self._message = msg
```

Comment:
Sure but can you make controller_test send and receive a message, just as a sanity check that everything works when multiple tests are doing so?